sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+!1 = !{!3, !4}            ; The first kernel parameter has two decorations.
+!2 = !{}                  ; The second kernel parameter has no decorations.
+!3 = !{i32 7742}          ; This is the integer value of the first decoration.
+!4 = !{i32 7743, i32 10}  ; The first operand is the integer value of the
+                          ;   second decoration.  Additional operands are
+                          ;   "extra operands" to the decoration.  These
+                          ;   operands may be either integer literals or string
+                          ;   literals.
+```
+
+### Property on a global variable
+
+When a property on a global variable needs to be represented in SPIR-V, we
+generally translate the property into a SPIR-V **OpDecorate** instruction for
+the corresponding module scope (global) **OpVariable**.  Again, there is no
+existing mechanism to do this in the SPIR-V LLVM Translator, so we propose the
+following mechanism.
+
+An LLVM IR global variable definition may optionally have a metadata kind of
+`!spirv.Decorations`.  If it does, that metadata node has one operand for each
+of the global variable's decorations.  To illustrate:
+
+```
+@MyVariable = global %MyClass !spirv.Decorations !0
+!0 = !{!1, !2}            ; Each operand in this metadata represents one
+                          ;   decoration on the variable.
+!1 = !{i32 7744}          ; This is the integer value of the first decoration.
+!2 = !{i32 7745, i32 20}  ; The first operand is the integer value of the
+                          ;   second decoration.  Additional operands are
+                          ;   "extra operands" to the decoration.  These
+                          ;   operands may be either integer literals or string
+                          ;   literals.
+```
+
+### Property on a structure member of a non-global variable
+
+As we noted earlier, a property on a structure member variable is represented
+in LLVM IR as a call to the intrinsic function `@llvm.ptr.annotation`, where
+the annotation string starts with the prefix `"sycl-properties:"`.  In order to
+understand how these SYCL properties are translated into SPIR-V, it's useful to
+review how a normal (i.e. non-SYCL) call to `@llvm.ptr.annotation` is
+translated.
+
+The existing behavior of the SPIR-V LLVM Translator is to translate this call
+into one (or both) of the following:
+
+* An **OpDecorate** instruction that decorates the intermediate pointer value
+  that is returned by the intrinsic (i.e. the pointer to the member variable).
+
+* An **OpMemberDecorate** instruction that decorates the member variable
+  itself.
+
+In both cases, the decoration is a single **UserSemantic** decoration where the
+string literal is the same as the string literal in the LLVM annotation.
+
+When a SYCL structure member property needs to be represented in SPIR-V,
+however, we prefer to represent each property as an extended SPIR-V decoration
```

Comment:
@gmlueck by 'extended SPIR-V decoration' you mean, some sort of non-semantic decoration with variadic arguments?

Reply:
No.  I mean that we would create a SPIR-V extension that defines new "Decoration" IDs that we would use with `OpDecorate`.  As an example of such an extension, see:

https://github.com/intel/llvm/blob/ff903a312d98d576cb3c443f26d71c8e942b7ec6/sycl/doc/extensions/DeviceGlobal/SPV_INTEL_global_variable_decorations.asciidoc

Reply:
Got it, thanks!